src/EditorFeatures/Core/ExtractMethod/ExtractMethodCommandHandler.cs

```diff
@@ -224,55 +184,76 @@ public bool ExecuteCommand(ExtractMethodCommandArgs args, CommandExecutionContex
+            Document document, TextSpan span, ExtractMethodGenerationOptions options, ExtractMethodResult result, CancellationToken cancellationToken)
         {
-            // We are about to show a modal UI dialog so we should take over the command execution
-            // wait context. That means the command system won't attempt to show its own wait dialog 
-            // and also will take it into consideration when measuring command handling duration.
-            var project = document.Project;
-            var solution = project.Solution;
-            var notificationService = solution.Services.GetService<INotificationService>();
-
-            // see whether we will allow best effort extraction and if it is possible.
-            if (!_globalOptions.GetOption(ExtractMethodPresentationOptionsStorage.AllowBestEffort, document.Project.Language) ||
-                !result.Status.HasBestEffort() ||
-                result.DocumentWithoutFinalFormatting == null)
+            // If we succeeded without any problems, just proceed without notifying the user.
+            if (result is { Succeeded: true, Reasons.Length: 0, DocumentWithoutFinalFormatting: not null })
+                return result;
+
+            // We have some sort of issue.  See what the user wants to do.  If we have no way to inform the user bail
+            // out rather than doing something wrong.
+            var notificationService = document.Project.Solution.Services.GetService<INotificationService>();
+            if (notificationService is null)
+                return null;
+
+            // The initial extract method failed, or it succeeded with warning reasons.  Attempt again with
+            // different options to see if we get better results.
+            var alternativeResult = await TryWithoutMakingValueTypesRefAsync(
+                document, span, result, options, cancellationToken).ConfigureAwait(false);
+            if (alternativeResult is { Succeeded: true, Reasons.Length: 0, DocumentWithoutFinalFormatting: not null })
             {
-                if (notificationService != null)
-                {
-                    await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
-                    notificationService.SendNotification(
-                        EditorFeaturesResources.Extract_method_encountered_the_following_issues + Environment.NewLine +
-                        string.Join("", result.Reasons.Select(r => Environment.NewLine + "  " + r)),
+                // We are about to show a modal UI dialog so we should take over the command execution
+                // wait context. That means the command system won't attempt to show its own wait dialog 
+                // and also will take it into consideration when measuring command handling duration.
+                await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
```

Comment:
>await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken); [](http://example.com/codeflow?start=16&length=87)

nit: looks like all codepaths from here require being on the main thread. Could just do this after the TryWithoutMakingValueTypesRefAsync call.

Reply:
can do!

Reply:
this was taken care of in followup.